### PR TITLE
NIFI-3281 - fix for (S)FTP processors when using EL against FFs

### PR DESCRIPTION
**Pull request by pvillard31**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FTPTransfer.java [hunk 1]
```diff
@@ -304,6 +304,12 @@ public void flush() throws IOException {
     }
 
     @Override
+    public void flush(final FlowFile flowFile) throws IOException {
+        final FTPClient client = getClient(flowFile);
+        client.completePendingCommand();
```
Comment (m-hogue) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FTPTransfer.java]: Should we handle the case where `client.completePendingCommand()` returns false? Per the javadocs [1], it returns false if the command couldn't be completed. 

[1] https://commons.apache.org/proper/commons-net/apidocs/org/apache/commons/net/ftp/FTPClient.html#completePendingCommand()

Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FTPTransfer.java]: or at least in this methods case return that boolean and explain the same meaning so the caller can decide case by case how/if to handle.  Good catch

**Comment by pvillard31** [comment]
Thanks @m-hogue and @joewitt for your feedback, I just pushed a commit that should address your comments: retrieve the boolean returned by ``completePendingCommand()`` and, if false, send the FF to ``REL_COMMS_FAILURE`` relationship. Also added a unit test for ``ListFTP`` to check attributes being written on flow files.

**Comment by m-hogue** [comment]
LGTM. Thanks!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java [hunk 2]
```diff
@@ -94,7 +94,7 @@
 
     @Override
     protected String getPath(final ProcessContext context) {
-        return context.getProperty(REMOTE_PATH).getValue();
+        return context.getProperty(REMOTE_PATH).evaluateAttributeExpressions().getValue();
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java]: I think we should pass FlowFile to evaluate EL for REMOTE_PATH, too.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FileTransfer.java [hunk 3]
```diff
@@ -39,6 +39,8 @@
 
     void flush() throws IOException;
 
+    boolean flush(FlowFile flowFile) throws IOException;
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FileTransfer.java]: Similar to `flush()`, other methods those use `getClient` (in FTPTransfer) or `getChannel` (in SFTPTransfer) should pass incoming FlowFile in order to evaluate EL properly. [NIFI-3590](https://issues.apache.org/jira/browse/NIFI-3590) is essentially the same to NIFI-3281 which is caused by not passing FlowFile.

Do we want to address it, too? Ideally, the codes `getClient(null)` and `getChannel(null)` should be removed completely.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java (hunk 2 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java]: ``ListFileTransfer`` is only used for ``ListFTP`` and ``ListSFTP`` and they do not accept incoming flow files. There is no flow file to use when evaluating the property.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FileTransfer.java (hunk 3 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/FileTransfer.java]: Thanks @ijokarumawak, will address asap in this PR.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java (hunk 2 above)
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFileTransfer.java]: You're right, I was thinking it was FileTransfer. Thanks!

**Comment by pvillard31** [comment]
@ijokarumawak - just pushed another commit to use flow file wherever possible and added a unit test to confirm this solves NIFI-3590 as well. It remains few ``getClient(null)`` and ``getChannel(null)`` but that's only when there is no input allowed for the processor (listing operation).

**Comment by ijokarumawak** [comment]
@pvillard31 Sorry for the delay. Thanks for including the fixes for NIFI-3590. I've tested Get/Put/List/Fetch/FTP/SFTP processors with EL, especially Put and Fetch with EL using incoming FlowFile attributes. All worked well. NiFi flow template that I used for testing is available here:
https://gist.github.com/ijokarumawak/b347520f6a7b05beca3ed8efe74d015e

LGTM, +1. Merging into master. Thanks!

**Pull request closed without merging**